### PR TITLE
Propose a new architecture with focus on scalability and robustness

### LINKED ISSUE #315: [Discussion] Switch to use Job API with ArrayJob semantics
From Aldo Culquicondor:
> I left you a message in k8s Slack, but posting it here too in case you are not so active there.
>
>I'm a sig-scheduling maintainer and I'm now looking into supporting array Jobs https://github.com/kubernetes/kubernetes/issues/97169
>We want this to enable MPI workloads on the Job API. I noticed that the MPI operator uses a StatefulSet. That seems like the best idea today. I'm wondering what it would take to migrate to Job, other than having a stable host name.
>
>For extra context, we are also looking into having Job level orchestration for provisioning and scheduling. We have published this doc on how we envision it to work: bit.ly/k8s-job-management
>
>I hope these topics interest you. Feel free to answer here or on Slack,
>

Related issue: https://github.com/kubernetes/kubernetes/issues/97169

cc @alculquicondor @ahg-g @kubeflow/wg-training-leads 

Let's use this issue to discuss with the community.

### PR DESCRIPTION
**Pull request by alculquicondor**
Fixes #315

### CODE REVIEW

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [inline proposals/scalable-robust-operator.md:121]:
```suggestion
  - The controller generates a single key and share it with the launcher and
```

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [inline proposals/scalable-robust-operator.md:126]:
```suggestion
  - When encrypted communication is not a requirement, users have the
```

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [inline proposals/scalable-robust-operator.md:147]: Add: because workers connect back to the launcher.

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [inline proposals/scalable-robust-operator.md:107]: the scale issue here is that to workaround the security concern, each job needs its own namespace, which is not scalable.

File: proposals/scalable-robust-operator.md
Comment (carmark) [inline proposals/scalable-robust-operator.md:155]: Do u mean the init-container `kubectl-delivery` will be removed? If that, how could we  keep the launcher start after the  workers?

File: proposals/scalable-robust-operator.md
Comment (carmark) [inline proposals/scalable-robust-operator.md:107]: The launcher has only the execution permission on the worker pods belong it's job, you can find it at [code](https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/v1/mpi_job_controller.go#L1222)

File: proposals/scalable-robust-operator.md
Comment (zw0610) [inline proposals/scalable-robust-operator.md:94]: I believe this issue is addressed in v1 controller when we enable horovod elastic work on mpi-operator. https://github.com/kubeflow/mpi-operator/blob/master/proposals/elastic-horovod.md

Besides, I'm not sure what it means 'workers belonging to the same or different jobs'. What is the user case to increase the number of workers for a *different* job?

File: proposals/scalable-robust-operator.md
Comment (zw0610) [inline proposals/scalable-robust-operator.md:108]: What are user cases for launcher pod retry?

File: proposals/scalable-robust-operator.md
Comment (zw0610) [inline proposals/scalable-robust-operator.md:104]:
```suggestion
    and has to be paid for every job. The API calls also cause additional
```

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [reply proposals/scalable-robust-operator.md:108]: Running jobs on a spot/preemptible VMs.

Comment (zw0610) [reply proposals/scalable-robust-operator.md:108]: In my perspective, I don't think the launcher pod should be scheduled to a spot/preemptible VMs. For the worker pods, yes, because worker pods are stateless. However, the launcher pod seems a stateful instance, whose re-restart means a new job.

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [reply proposals/scalable-robust-operator.md:94]: The sentence perhaps needs some rephrasing, I think the point here is that the current setup doesn't scale when increasing the number of workers of a job (e.g., running a 1000 worker pod Job), or increasing the number of jobs (running 100s of MPI jobs). 

The current design puts a lot pressure on central component, the api server:

- `kubectl exec` runs through the API server, and so all jobs running on the same cluster will have to launch their workers using connections established through the api server, which is not scalable. This doesn't scale as we increase the total number of workers across all jobs running on the same cluster.
- pod cache syncs are expensive on large clusters, doing that for each job results in streaming a large amount of state from the api server to the launcher pod. This doesn't scale as we increase the number of jobs.

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:155]: There are 2 options:
1. Do nothing, leave such handling to the gang scheduler. For this reason it is important that the launcher can do retries.
1. To not create the launcher until all the worker pods are running. So everything is handled by the controller, no need for extra cache syncs.

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:107]: Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:108]: Still, when running the workers in a preemptible VM, if any of them fail, the entire job fails, including the launcher. So we need launcher retries.

**Review by alculquicondor — COMMENTED** [review]
Thanks for your comments so far. Please let me know if I missed anything else in the background. I got all the details by reading code.

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md]: According to the other comment, this is not a security issue, as the role has the list of pods. But it's still a scalability issue. See update.

File: proposals/scalable-robust-operator.md
Comment (terrytangyuan) [reply proposals/scalable-robust-operator.md]: @alculquicondor What MPI implementation are you expecting users to use after launcher gets restarted to make sure the workers are aware and can reconnect to the new launcher?

File: proposals/scalable-robust-operator.md
Comment (terrytangyuan) [inline proposals/scalable-robust-operator.md:168]: Could you add a link to "Indexed Jobs"?

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md]: The process goes like this:
1. a worker terminates unexpectedly
2. the launcher notices the worker failure, it closes the rest of ssh connections and terminates with a failure
3. the launcher restarts, launches new ssh connections to the workers
This is independent of the MPI implementation.

File: proposals/scalable-robust-operator.md
Comment (carmark) [reply proposals/scalable-robust-operator.md:160]: I don't think so, you would like to speed up startup time, but it does not if you want the launcher to retry for failures. And the scheduler should not know about the job startup strategy.

Comment (carmark) [reply proposals/scalable-robust-operator.md:160]: By the same time, how do you verify the  real launcher failed reason for option 1? for example,
* workers does not startup, the launcher failed and restart
* training job failed, the launcher failed.

File: proposals/scalable-robust-operator.md
Comment (zw0610) [reply proposals/scalable-robust-operator.md]: > Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

It does. We introduced this feature in supporting Horovod Elastic: https://github.com/kubeflow/mpi-operator/pull/334

File: proposals/scalable-robust-operator.md
Comment (zw0610) [reply proposals/scalable-robust-operator.md]: > Still, when running the workers in a preemptible VM, if any of them fail, the entire job fails, including the launcher. So we need launcher retries.

From what I observed from other HPC system, like Slurm, for jobs that are not fault-tolerant, when some of the worker fails, it does not retry via a launcher restart. Instead, the entire job will be marked as failed with it resources released. The system will re-queue the failed job if 'retry' is demanded by the user and create a new job. Such process is compatible with the contemporary design of mpi-operator.

File: proposals/scalable-robust-operator.md
Comment (gaocegege) [reply proposals/scalable-robust-operator.md]: If we replace kubectl exec with ssh, we can also support the elastic feature, thus I do not think it is a problem for us if I understand it correctly.

File: proposals/scalable-robust-operator.md
Comment (gaocegege) [inline proposals/scalable-robust-operator.md:250]: I am wondering how to support elastic mode with statefulset? We may add/remove workers in the fly.

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:160]: What do you mean by "job startup strategy"?

As for differentiating failures, I don't think we can or should do that. There is another type of failure which is kind of a mix of the two you described: a worker goes down. In this case, the job might have already started.

But perhaps option 2 is reasonable. WDYT?

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:250]: You just change `.spec.replicas`.

File: proposals/scalable-robust-operator.md
Comment (alculquicondor) [reply proposals/scalable-robust-operator.md]: OTOH, it kind of goes against the declarative and fault-tolerant approach of k8s APIs, including the Job API. Retries is what kubernetes users would expect. And if they don't need it, they could always disable it.

**Comment by alculquicondor** [comment]
There are 2 main open discussions pending:
- Should we hold the creation of the launcher until the workers are all ready? https://github.com/kubeflow/mpi-operator/pull/360#discussion_r634211580
- Are retries necessary? https://github.com/kubeflow/mpi-operator/pull/360#discussion_r634222943

Please leave your thoughts

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [reply proposals/scalable-robust-operator.md]: I think it is useful if we make automatic retries an option that we default to 0.

File: proposals/scalable-robust-operator.md
Comment (ahg-g) [reply proposals/scalable-robust-operator.md:160]: If we want to support retries, then by product we want option 1, correct?

Comment (alculquicondor) [reply proposals/scalable-robust-operator.md:160]: we can still use option 2. It might make startups less flaky. But it's not a guarantee anyways. A worker can simply just fail in between the time it reported running and the launcher started running.

**Comment by alculquicondor** [comment]
I asked for 10 min in the next AutoML and Trainging WG meeting. Would you be able to make it @terrytangyuan @gaocegege ?

**Comment by terrytangyuan** [comment]
I cannot make it due to scheduling conflicts. In the meantime, I'd encourage all existing reviewers to check if your comments can be resolved or discuss if there are further concerns on the proposal (which might be more efficient as we are located in more than 4 timezones).

**Comment by ahg-g** [comment]
> I cannot make it due to scheduling conflicts. In the meantime, I'd encourage all existing reviewers to check if your comments can be resolved or discuss if there are further concerns on the proposal (which might be more efficient as we are located in more than 4 timezones).

+1. Can we also assign an "approver"?

**Comment by terrytangyuan** [comment]
We should try to reach consensus from people listed in [OWNERS](https://github.com/kubeflow/mpi-operator/blob/master/OWNERS). People from @kubeflow/wg-training-leads can also approve after that.

**Comment by alculquicondor** [comment]
@rongou any thoughts?

**Comment by gaocegege** [comment]
@carmark @zw0610 @Jeffwan PTAL thanks.

**Comment by alculquicondor** [comment]
One topic that was raised at today's meeting was where to place the code. This question is important because there are breaking changes and the new controller wouldn't be able to properly process an existing job.

I was thinking that we can just modify v1 code, as it hasn't been released, AFAIK. Is this not the case? Should we create a new folder? like `v1-ssh` or something like that? My worry is that this would become a maintenance problem, as now changes would have to be applied in 2 places. And long term, the existing v1 code should be removed. But if v1 was never released, it should be safe to just replace it.

**Comment by terrytangyuan** [comment]
It's true that v1 is not officially released yet but I believe there are several companies that are already running it in production which might be concerning.

**Comment by alculquicondor** [comment]
I see. Then maybe we can create the new controller in a entire new module, so that we can start with fresh dependencies and not be blocked by kube-batch's (see #364)

> We should try to reach consensus from people listed in OWNERS. People from @kubeflow/wg-training-leads can also approve after that.

If they don't respond, can we assume lazy consensus?

**Comment by zw0610** [comment]
> It's true that v1 is not officially released yet but I believe there are several companies that are already running it in production which might be concerning.

I believe the API version `v1`, `v1alpha1` and `v1alpha2` is not necessary bounded to the controller version, which means, if this proposal wish to add a kind of new controller to the `v1` API, all the contributors need to do is add the implementation under `pkg/controllers/v1` but in a different file like `pkg/controllers/v1/mpi_job_alternative_controller.go`. After the implementation is done, you can add option in `cmd/mpi-operator.v1` to let user choose which controller to use. In this way, we can avoid users who are already using the `v1` API can the existing controller implementation.

**Comment by alculquicondor** [comment]
The problem is not necessarily the API. Let's say someone has running jobs with the existing v1 controller. If they upgrade to the new proposed controller, the existing jobs would have orphan resources that the new controller won't manage. And the existing workers wouldn't work, because they don't have stable hostnames.

**Comment by alculquicondor** [comment]
In any case, I think the decision of where to place the code comes later. Can we have an agreement on the direction of the new architecture? See https://github.com/kubeflow/mpi-operator/pull/360#issuecomment-845445542

**Comment by alculquicondor** [comment]
Thanks, I'll work on that throughout the week.

**Comment by alculquicondor** [comment]
I ran some experiments with these characteristics:

- 21 nodes
- 600 running pods (not part of the job)
- Job with 3 slots per worker
- All nodes already have the image (so no image pull time involved)

Then I run two jobs, one with 3 workers and one with 20 workers. Each task is doing:
1. print its rank
2. Execute a simple calculation
3. Do `MPI_Reduce` targeting rank 0

So really, most of the time is spent on setup, but some communication is being tested too.

For each run, I recorded the Start time and Completion time as reported by the mpi-operator controller.

For 3 workers, I got an average of 9s for the current operator and 3s for the ssh proposal.
For 20 workers, I got an average of 19s for the current operator and 11s for the ssh proposal.

All runs here:

image

You can see an important improvement. I used a reasonably small cluster and somewhat small jobs. Let me know if you would like any more details about my quick experiments.

However, let me reiterate on theoretical analysis already presented, which I think is important on its own. The apiserver is a critical component in the cluster. We should avoid stressing it with requests that intend to do pod-to-pod communication. The apiserver has limits for ongoing open connections and it throttles API requests. With the current architecture, just a few dozens of jobs could cause starvation to API requests from critical components such as the scheduler or the controller manager.

**Comment by alculquicondor** [comment]
@terrytangyuan @zw0610 anything to add?

**Comment by terrytangyuan** [comment]
Thanks for putting this together! I think both theoretical and analytical improvements look great to me. 

As it's already close to the weekend, **please leave additional comments by June 16th and I will merge this if there's no major concerns by then.**

**Comment by johnugeorge** [comment]
I feel that this proposal is a strong case for a revamp based on https://github.com/kubeflow/mpi-operator/pull/360#issuecomment-861632258. 

/lgtm

**Comment by gaocegege** [comment]
/lgtm

Thanks for the proposal.

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**
